packages/lib/src/__tests__/basic-io.test.ts

```diff
@@ -0,0 +1,189 @@
+import { CLIError } from '@oclif/errors'
+
+import { inputAndOutputItem, Naming, outputItem, outputList, Sorting } from '../basic-io'
+import * as format from '../format'
+import { InputProcessor, UserInputProcessor } from '../input'
+import * as inputBuilder from '../input-builder'
+import { IOFormat } from '../io-util'
+import * as output from '../output'
+import * as outputBuilder from '../output-builder'
+import { SmartThingsCommandInterface } from '../smartthings-command'
+import { buildMockCommand } from './test-lib/mock-command'
+import { SimpleType } from './test-lib/simple-type'
+
+
+describe('basic-io', () => {
+	const item = { 'str': 'string', num: 5 }
+	const list = [item]
+	let baseCommand: SmartThingsCommandInterface
+	let formatAndWriteItemSpy: jest.SpyInstance<Promise<void>, [command: (SmartThingsCommandInterface & format.TableCommonOutputProducer<unknown>) | (SmartThingsCommandInterface & format.CustomCommonOutputProducer<unknown>), item: unknown, defaultIOFormat?: IOFormat | undefined]>
+	let formatAndWriteListSpy: jest.SpyInstance<Promise<void>, [command: (SmartThingsCommandInterface & Sorting & Naming) | (SmartThingsCommandInterface & format.TableCommonListOutputProducer<unknown> & Naming) | (SmartThingsCommandInterface & format.CustomCommonListOutputProducer<unknown> & Naming), list: unknown[], includeIndex?: boolean | undefined]>
+
+	beforeEach(() => {
+		baseCommand = {
+			...buildMockCommand(),
+			flags: {
+				output: 'output.yaml',
+			},
+		}
+		formatAndWriteItemSpy = jest.spyOn(format, 'formatAndWriteItem')
+		formatAndWriteItemSpy.mockImplementation(async () => { /* empty */})
+
+		formatAndWriteListSpy = jest.spyOn(format, 'formatAndWriteList')
+		formatAndWriteListSpy.mockImplementation(async () => { /* empty */})
```

Comment:
I believe the spy declarations could be moved out of beforeEach as they should be managed by the call to `clearAllMocks` below.

Reply:
They could be but I put them here to reduce duplication. There is so much duplication in these tests that I haven't really figured out how to clean up properly.

Reply:
I meant to say, move them up into the `describe` context where they are initially declared. This will achieve the reusability while also leveraging the call to `clearAllMocks` after each test to clear state.

Reply:
Well that works out because based on our conversation above, that's what I ended up doing. 😄